pvlib/iotools/psm4.py

```diff
@@ -0,0 +1,819 @@
+"""
+Functions for reading and retrieving data from NSRDB PSM4. See:
+https://developer.nrel.gov/docs/solar/nsrdb/nsrdb-GOES-aggregated-v4-0-0-download/
+https://developer.nrel.gov/docs/solar/nsrdb/nsrdb-GOES-tmy-v4-0-0-download/
+https://developer.nrel.gov/docs/solar/nsrdb/nsrdb-GOES-conus-v4-0-0-download/
+https://developer.nrel.gov/docs/solar/nsrdb/nsrdb-GOES-full-disc-v4-0-0-download/
+"""
+
+import io
+from urllib.parse import urljoin
+import requests
+import pandas as pd
+from json import JSONDecodeError
+
+NSRDB_API_BASE = "https://developer.nrel.gov/api/nsrdb/v2/solar/"
+PSM4_AGG_ENDPOINT = "nsrdb-GOES-aggregated-v4-0-0-download.csv"
+PSM4_TMY_ENDPOINT = "nsrdb-GOES-tmy-v4-0-0-download.csv"
+PSM4_CON_ENDPOINT = "nsrdb-GOES-conus-v4-0-0-download.csv"
+PSM4_FUL_ENDPOINT = "nsrdb-GOES-full-disc-v4-0-0-download.csv"
+PSM4_AGG_URL = urljoin(NSRDB_API_BASE, PSM4_AGG_ENDPOINT)
+PSM4_TMY_URL = urljoin(NSRDB_API_BASE, PSM4_TMY_ENDPOINT)
+PSM4_CON_URL = urljoin(NSRDB_API_BASE, PSM4_CON_ENDPOINT)
+PSM4_FUL_URL = urljoin(NSRDB_API_BASE, PSM4_FUL_ENDPOINT)
+
+PARAMETERS = (
+    'air_temperature', 'dew_point', 'dhi', 'dni', 'ghi', 'surface_albedo',
+    'surface_pressure', 'wind_direction', 'wind_speed')
+PVLIB_PYTHON = 'pvlib python'
+
+# Dictionary mapping PSM4 response names to pvlib names
+VARIABLE_MAP = {
+    'GHI': 'ghi',
+    'DHI': 'dhi',
+    'DNI': 'dni',
+    'Clearsky GHI': 'ghi_clear',
+    'Clearsky DHI': 'dhi_clear',
+    'Clearsky DNI': 'dni_clear',
+    'Solar Zenith Angle': 'solar_zenith',
+    'Temperature': 'temp_air',
+    'Dew Point': 'temp_dew',
+    'Relative Humidity': 'relative_humidity',
+    'Pressure': 'pressure',
+    'Wind Speed': 'wind_speed',
+    'Wind Direction': 'wind_direction',
+    'Surface Albedo': 'albedo',
+    'Precipitable Water': 'precipitable_water',
+    'AOD': 'aod',
+}
+
+# Dictionary mapping pvlib names to PSM4 request names
+# Note, PSM4 uses different names for the same variables in the
+# response and the request
+REQUEST_VARIABLE_MAP = {
+    'ghi': 'ghi',
+    'dhi': 'dhi',
+    'dni': 'dni',
+    'ghi_clear': 'clearsky_ghi',
+    'dhi_clear': 'clearsky_dhi',
+    'dni_clear': 'clearsky_dni',
+    'solar_zenith': 'solar_zenith_angle',
+    'temp_air': 'air_temperature',
+    'temp_dew': 'dew_point',
+    'relative_humidity': 'relative_humidity',
+    'pressure': 'surface_pressure',
+    'wind_speed': 'wind_speed',
+    'wind_direction': 'wind_direction',
+    'albedo': 'surface_albedo',
+    'precipitable_water': 'total_precipitable_water',
+    'aod': 'aod',
+}
+
+
+def get_nsrdb_psm4_aggregated(latitude, longitude, api_key, email,
+                              year='2023', time_step=60,
```

Comment:
I just noticed the default value `year='2023'` in this function and some others.  Is there a good reason for it?  I don't see much convenience in having a default year since presumably most users will override it for the year they actually need anyway.  And if it's possible to leave the parameter unspecified, it might funnel careless users towards requesting the wrong year unintentionally.

I think we should change it so that `year` does not have a default value for the calendar year (i.e. non-TMY) functions. @williamhobbs @adamrjensen any thoughts?

Reply:
Does the NSRDB return an informative error if a year is requested for which there is no data? That might be why a default was included.